shared/typeflow/codeql/typeflow/internal/TypeFlowImpl.qll

```diff
@@ -0,0 +1,456 @@
+    not irrelevantBound(n, t)
+  }
+
+  predicate bestTypeFlow(TypeFlowNode n, Type t, boolean exact) {
+    exactType(n, t) and exact = true
+    or
+    not exactType(n, _) and bestTypeFlow(n, t) and exact = false
+  }
+
+  private predicate bestTypeFlowOrTypeFlowBase(TypeFlowNode n, Type t, boolean exact) {
+    bestTypeFlow(n, t, exact)
+    or
+    typeFlowBase(n, t) and
+    exact = false and
+    not bestTypeFlow(n, _, _)
+  }
+
+  /**
+   * Holds if `n` has type `t` and this information is not propagated as a
+   * universal bound to a subsequent node, such that `t` might form the basis for
+   * a union type bound for that node.
+   */
+  private predicate unionTypeFlowBaseCand(TypeFlowNode n, Type t, boolean exact) {
+    exists(TypeFlowNode next |
+      joinStep(n, next) and
+      bestTypeFlowOrTypeFlowBase(n, t, exact) and
+      not bestTypeFlowOrTypeFlowBase(next, t, exact) and
+      not exactType(next, _)
+    )
+  }
+
+  private module HasUnionTypePropagation implements TypePropagation {
+    class Typ = Unit;
+
+    predicate candType(TypeFlowNode mid, Unit unit) {
+      exists(unit) and
+      (unionTypeFlowBaseCand(mid, _, _) or hasUnionTypeFlow(mid))
+    }
+
+    predicate supportsType = candType/2;
+  }
+
+  /**
+   * Holds if all incoming type flow can be traced back to a
+   * `unionTypeFlowBaseCand`, such that we can compute a union type bound for `n`.
+   * Disregards nodes for which we have an exact bound.
+   */
+  private predicate hasUnionTypeFlow(TypeFlowNode n) {
+    not exactType(n, _) and
+    (
+      // Optimized version of
+      // `forex(TypeFlowNode mid | joinStep(mid, n) | unionTypeFlowBaseCand(mid, _, _) or hasUnionTypeFlow(mid))`
+      ForAll<TypeFlowNode, RankedJoinStep, HasUnionTypePropagation>::flowJoin(n, _)
+      or
+      exists(TypeFlowScc scc |
+        sccRepr(n, scc) and
+        // Optimized version of
+        // `forex(TypeFlowNode mid | sccJoinStep(mid, scc) | unionTypeFlowBaseCand(mid, _, _) or hasUnionTypeFlow(mid))`
+        ForAll<TypeFlowScc, RankedSccJoinStep, HasUnionTypePropagation>::flowJoin(scc, _)
+      )
+      or
+      exists(TypeFlowNode mid | step(mid, n) and hasUnionTypeFlow(mid))
+      or
+      instanceofDisjunctionGuarded(n, _)
+    )
+  }
+
+  pragma[nomagic]
+  private Type getTypeBound(TypeFlowNode n) {
+    bestTypeFlow(n, result)
+    or
+    not bestTypeFlow(n, _) and result = n.getType()
+  }
+
+  pragma[nomagic]
+  private predicate unionTypeFlow0(TypeFlowNode n, Type t, boolean exact) {
+    hasUnionTypeFlow(n) and
+    (
+      exists(TypeFlowNode mid | anyStep(mid, n) |
+        unionTypeFlowBaseCand(mid, t, exact) or unionTypeFlow(mid, t, exact)
+      )
+      or
+      instanceofDisjunctionGuarded(n, t) and exact = false
+    )
+  }
+
+  /**
+   * Holds if there is a common (reflexive, transitive) subtype of the erased
+   * types `t1` and `t2`.
+   */
+  private predicate erasedHaveIntersection(Type t1, Type t2) {
```

Comment:
This predicate is somewhat volatile performance-wise, so we should probably just copy its implementation more verbatim. At the very least it needs to be nomagic'ed, and if we extract the `SrcRefType` in the same way as for Java, then we'll be more likely to get the two implementations unified by the compiler.

Reply:
Hmm... Doesn't look like it was unified:
```
EVALUATE NONRECURSIVE RELATION:
  `TypeFlowImpl::TypeFlow<Location::Location,TypeFlow::Input>::erasedHaveIntersection/2#bb495610`(entity t1, entity t2) :- 
    SENTINEL `Type::RefType.getSourceDeclaration/0#dispred#770ab75c`
    SENTINEL `project#TypeFlow::Input::getErasure/1#ad6f3159`
    {2} r1 = REWRITE `Type::RefType.getSourceDeclaration/0#dispred#770ab75c` WITH TEST InOut.0 't2' = InOut.1

    {2} r2 = SCAN r1 OUTPUT In.0 't2', In.0 't2'
    {2}    | JOIN WITH `project#TypeFlow::Input::getErasure/1#ad6f3159` ON FIRST 1 OUTPUT Lhs.1 't2', Lhs.0 't1'

    {1} r3 = SCAN r1 OUTPUT In.0 't2'
    {2}    | JOIN WITH `#Type::RefType.getASourceSupertype/0#dispred#418e5974Plus` ON FIRST 1 OUTPUT Rhs.1 't1', Lhs.0 't2'
    {2}    | JOIN WITH `project#TypeFlow::Input::getErasure/1#ad6f3159` ON FIRST 1 OUTPUT Lhs.1 't2', Lhs.0 't1'

    {2} r4 = r2 UNION r3

    {2} r5 = JOIN r4 WITH `project#TypeFlow::Input::getErasure/1#ad6f3159` ON FIRST 1 OUTPUT Lhs.1 't1', Lhs.0 't2'

    {2} r6 = JOIN r4 WITH `#Type::RefType.getASourceSupertype/0#dispred#418e5974Plus` ON FIRST 1 OUTPUT Rhs.1 't1', Lhs.1 't1'
    {2}    | JOIN WITH `project#TypeFlow::Input::getErasure/1#ad6f3159` ON FIRST 1 OUTPUT Lhs.1 't1', Lhs.0 't2'

    {2} r7 = r5 UNION r6
    return r7
```